core/src/main/java/brooklyn/location/access/PortForwardManagerImpl.java

```diff
@@ -160,11 +162,22 @@ protected int getNextPort() {
     @Override
     public void associate(String publicIpId, HostAndPort publicEndpoint, Location l, int privatePort) {
         associateImpl(publicIpId, publicEndpoint, l, privatePort);
+        emitAssociationEvent(publicIpId, publicEndpoint, l, privatePort);
     }
 
     @Override
     public void associate(String publicIpId, HostAndPort publicEndpoint, int privatePort) {
         associateImpl(publicIpId, publicEndpoint, null, privatePort);
+        emitAssociationEvent(publicIpId, publicEndpoint, null, privatePort);
+    }
+
+    private void emitAssociationEvent(String publicIpId, HostAndPort publicEndpoint, Location location, int privatePort) {
+        AssociationMetadata metadata = new AssociationMetadata(publicIpId, publicEndpoint, location, privatePort);
+        for (AssociationListener listener : associationListeners.keySet()) {
+            if (associationListeners.get(listener).apply(metadata)) {
```

Comment:
Looks dangerous in a multi-threaded world. You've separated the call of `keySet()` and `get(listener)`, so you've introduced the chance for another thread to remove the listener in between those two calls.

Better to do the following (thus allowing `ConcurrentHashMap` to give its guarantees):

    for (Map.Entry<AssociationListener, Predicate<? super AssociationMetadata> entry : associationListeners.entrySet()) {
        if (entry.getValue().apply(metadata)) {
            entry.getKey().onEvent(metadata);
        }
    }